src/app.ts

```diff
@@ -195,21 +200,71 @@ export const createAppTask = withSpid({
     });
   });
   withSpidApp.post("/introspect", async (req, res) => {
-    await redisGetTokenUser(`${SESSION_TOKEN_PREFIX}${req.body.token}`)
-      .mapLeft(err =>
-        res.json({
-          active: false,
-          error: err
-        })
+    // first check if token is blacklisted
+    await existsKeyTask(
+      redisClient,
+      `${SESSION_INVALIDATE_TOKEN_PREFIX}${req.body.token}`
+    )
+      .mapLeft(() => res.status(500).json("Cannot introspect token"))
+      .chain(
+        fromPredicate(
+          _ => _ === false,
+          () =>
+            res.status(403).json({
+              active: false
+            })
+        )
+      )
+      .chain(() =>
+        getTask(
+          redisClient,
+          `${SESSION_TOKEN_PREFIX}${req.body.token}`
+        ).mapLeft(() => res.status(500).json("Error while retrieving token"))
+      )
+      .chain<TokenUser>(maybeToken =>
+        maybeToken
+          .foldL(
+            () =>
+              // tslint:disable-next-line: no-any
+              fromLeft(res.status(404).json("Token not found")),
+            rawToken =>
+              fromEither(
+                parseJSON(rawToken, err =>
+                  res.status(500).json({
+                    detail: String(err),
+                    error: "Error parsing token"
+                  })
+                )
+              )
+          )
+          .chain(_ =>
+            fromEither(TokenUser.decode(_)).mapLeft(errs =>
+              res.status(500).json({
+                detail: String(errorsToError(errs)),
+                error: "Error while decoding token"
+              })
+            )
+          )
       )
       .chain(
         fromPredicate(
           () => config.INCLUDE_SPID_USER_ON_INTROSPECTION,
-          () => res.json({ active: true })
+          () => res.status(200).json({ active: true })
         )
       )
       .map(tokenUser => ({ active: true, user: tokenUser }))
-      .fold(identity, _ => res.json(_))
+      .fold(identity, _ => res.status(200).json(_))
+      .run();
+  });
+
+  withSpidApp.post("/invalidate", async (req, res) => {
```

Comment:
This API should works on 2 different way if the token is JWT or opaque.
JWT needs to store on redis the invalidate token
opaque token we can delete the existing token

Reply:
👍